e2e/test/nodepool/nodepool_test.go

```diff
@@ -0,0 +1,381 @@
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	releasev1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/release/v1alpha1"
+	"github.com/giantswarm/apiextensions/pkg/label"
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/reference"
+	expcapzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	capiv1alpha3 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	expcapiv1alpha3 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/e2e/env"
+	"github.com/giantswarm/azure-operator/v4/e2e/setup"
+	key2 "github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+func Test_Nodepool(t *testing.T) {
+	err := nodepool.Test(context.Background())
+	if err != nil {
+		t.Fatalf("%#v", err)
+	}
+}
+
+type Config struct {
+	ClusterID  string
+	CtrlClient client.Client
+	Logger     micrologger.Logger
+	Provider   *Provider
+	NodePoolID string
+}
+
+type Nodepool struct {
+	clusterID  string
+	ctrlClient client.Client
+	logger     micrologger.Logger
+	provider   *Provider
+	nodePoolID string
+}
+
+func New(config Config) (*Nodepool, error) {
+	if config.ClusterID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ClusterID must not be empty", config)
+	}
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.NodePoolID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.NodePoolID must not be empty", config)
+	}
+	if config.Provider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Provider must not be empty", config)
+	}
+
+	s := &Nodepool{
+		clusterID:  config.ClusterID,
+		ctrlClient: config.CtrlClient,
+		logger:     config.Logger,
+		nodePoolID: config.NodePoolID,
+		provider:   config.Provider,
+	}
+
+	return s, nil
+}
+
+const LabelVmSize = "node.kubernetes.io/instance-type"
+
+func (s *Nodepool) Test(ctx context.Context) error {
+	clusterID := s.clusterID
+
+	setupNodePoolID := s.nodePoolID
+	setupNodePoolVmSize := env.AzureVMSize()
+	setupNodePoolReplicas := 2
+
+	newNodepoolID := "t3st"
+	newNodePoolVmSize := "Standard_D3_v2"
+	newNodePoolReplicas := 3
+
+	err := s.CreateNodePool(ctx, newNodepoolID, int32(newNodePoolReplicas), newNodePoolVmSize)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = s.WaitForNodesReady(ctx, newNodePoolReplicas+setupNodePoolReplicas)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = s.assertRightNumberOfNodes(ctx, setupNodePoolID, newNodepoolID, setupNodePoolReplicas, newNodePoolReplicas, setupNodePoolVmSize, newNodePoolVmSize)
```

Comment:
We assert that the two different node pools contain different number of nodes. Since normally node pools will have different virtual machine sizes, I'm selecting nodes by nodepool label or vm size label, and making sure there are the right number.